Add option to redirect trace output to file

fbs opened the pull request:
This adds the option to redirect all output generated by the "trace
program" to be output to a file as discussed in #605. All other output
generated by `bpftrace` will still be sent either `stdout` or `stderr`.

It's implementation is quite straight forward. The `BPFtrace` class
stores a reference to an `ostream` object which will be used for all
output, replacing `cout` everywhere.
The `printf` command now uses `snprintf` to print to a buffer which then
gets written to the output stream. As the size isn't known initially
`sprintf` is called twice, once to calculate the required buffer size
and once to fill the buffer. A simple micro benchmark that compares the
two by doing 1k rounds shows that the new `snprintf` loop less than 5x
slower when writing to `/dev/null`. However, when writing to `stdout` the
difference drops to about 1.5x, as `stdout` is "slow".

Instead of calling `system` the `system` builtin now uses `exec_system`
from as that captures `stdout`. `stderr` is not captured (yet).

Closes #605

ajor commented on src/bpftrace.cpp:
```diff
@@ -314,23 +346,24 @@ void perf_event_printer(void *cb_cookie, void *data, int size __attribute__((unu
     t = time(NULL);
     tmp = localtime(&t);
     if (tmp == NULL) {
+      std::cerr << "localtime: " << strerror(errno) << std::endl;
```
Won't this error get printed out twice now?

ajor commented on src/bpftrace.cpp:
```diff
@@ -409,36 +405,13 @@ void perf_event_printer(void *cb_cookie, void *data, int size __attribute__((unu
-  }
+  // Calculate required size by passing NULL
+  int len = format(NULL, 0, fmt, arg_values);
+  char * buffer = new char[len];
```
We don't want to be using `new`/`delete` - can this be changed to smart pointers?

ajor commented on src/bpftrace.cpp:
```diff
@@ -409,36 +405,13 @@ void perf_event_printer(void *cb_cookie, void *data, int size __attribute__((unu
-      abort();
-  }
+  // Calculate required size by passing NULL
+  int len = format(NULL, 0, fmt, arg_values);
```
If we have a small buffer on the stack (like for `system`) we could try writing to that and only go on to allocate a buffer on the heap if `format()` reports that the stack buffer isn't big enough

fbs replied:
Makes sense, could you take a look at the new version?

ajor commented on src/bpftrace.cpp:
```diff
@@ -1145,14 +1119,15 @@ int BPFtrace::print_map_hist(IMap &map, uint32_t top, uint32_t div)
 
-    std::cout << std::endl;
+    out_ << std::endl;
+
```
Extra newline

ajor commented on src/bpftrace.h:
```diff
@@ -51,10 +52,12 @@ inline DebugLevel operator++(DebugLevel& level, int)
 public:
-  BPFtrace() : ncpus_(get_possible_cpus().size()) { }
+  BPFtrace() : out_(std::cout),ncpus_(get_possible_cpus().size()) { }
+  BPFtrace(std::ostream& o) : out_(o),ncpus_(get_possible_cpus().size()) { }
```
Could we just do this using a single constructor with a default value for `o`? I don't like duplicating code (`get_possible_cpus().size()`) if we can help it

fbs replied:
done

ajor commented on src/bpftrace.h:
```diff
@@ -51,10 +52,12 @@ inline DebugLevel operator++(DebugLevel& level, int)
   virtual ~BPFtrace();
   virtual int add_probe(ast::Probe &p);
   int num_probes() const;
+  std::ostream& outputstream() { return out_; }
```
I think this function can be made `const`

fbs replied:
`const ostream` isn't possible, or did you mean something else?

ajor replied:
Sorry, I meant `std::ostream& outputstream() const {...}`, not 100% sure about this though

fbs replied:
This seems to work, the runtime tests pass.

ajor commented on src/bpftrace.cpp:
```diff
@@ -361,46 +394,9 @@ void perf_event_printer(void *cb_cookie, void *data, int size __attribute__((unu
-    (void)ret;
-
+    char buffer[255];
+    format(buffer, 255, fmt, arg_values);
```
I'm just thinking - would there be any problem with longer system calls like with printf?

Even if we do want to keep it limited, we shouldn't run the command if our string is truncated. Could get some nasty things happening.

fbs replied:
I couldn't find a reason but just left it as is.

Will do this:
```
int size = format(buffer, BUFSIZE, fmt, arg_values);
// Return value is required size excluding null byte
if (size >= BUFSIZE) {
  std::cerr << "syscall() command to long" << std::endl;
  abort();
}
```

ajor commented on src/bpftrace.cpp:
```diff
@@ -39,6 +39,37 @@ namespace bpftrace {
+        args.at(3)->value(), args.at(4)->value(), args.at(5)->value());
+      break;
+    default:
+      std::cerr << "printf() can only take up to 7 arguments (" << args.size() << ") provided" << std::endl;
```
```suggestion
      std::cerr << "format() can only take up to 7 arguments (" << args.size() << ") provided" << std::endl;
```
Old code was wrong too, but let's fix it

javierhonduco commented:
Could you please add some runtime tests and update the documentation once this is ready? :)

fbs commented on src/bpftrace.cpp:
```diff
@@ -314,23 +346,24 @@ void perf_event_printer(void *cb_cookie, void *data, int size __attribute__((unu
   }
   else if (printf_id == asyncactionint(AsyncAction::cat))
-  {
+    {
```
fixed

ajor commented on src/bpftrace.cpp:
```diff
@@ -383,46 +421,15 @@ void perf_event_printer(void *cb_cookie, void *data, int size __attribute__((unu
+    // Return value is required size EXCLUDING null byte
+    if (size >= BUFSIZE) {
+      std::cerr << "syscall() command to long (" << size << ") bytes" << std::endl;
+      abort();
```
Don't think this should be a fatal error. Just printing out the contents of buffer and the message above should be enough

fbs replied:
I'm happy with both, updated it

fbs commented:
@ajor  I've rebased onto master and added some tests. Can you take a final look? The `cat` & `rm` parts of the test feel a bit hacky. But imo it's better than adding a one-off command to the test suite.

ajor commented on tests/runtime/file-output:
```diff
@@ -0,0 +1,19 @@
+NAME printf to file
+RUN bpftrace -v -e 'i:ms:10 { printf("%s %d\n", "SUCCESS", 1); exit() }' -o "out.tmp" >/dev/null; cat "out.tmp"; rm "out.tmp"
```
I'd suggest writing to a file like `/tmp/bpftrace-file-output-test` to avoid  writing to the regular file system and reduce the chance of that file already existing

fbs replied:
woops

ajor commented on tests/runtime/file-output:
```diff
@@ -0,0 +1,19 @@
+TIMEOUT 5
+
+NAME print map to file
+RUN bpftrace -v -e 'i:ms:10 { @=lhist(50, 0, 100, 10); exit();}' -o "out.tmp" >/dev/null; cat "out.tmp"; rm "out.tmp"; rm "out.tmp"
```
Doubled up on the `rm` at the end

fbs replied:
fixed

ajor commented on tests/runtime/file-output:
```diff
@@ -0,0 +1,19 @@
+NAME printf to file
+RUN bpftrace -v -e 'i:ms:10 { printf("%s %d\n", "SUCCESS", 1); exit() }' -o bpftrace-file-output-test >/dev/null; cat bpftrace-file-output-test; rm bpftrace-file-output-test
```
I also think it's important to put this file we're writing to in `/tmp` so that it gets auto cleaned up eventually if our tests fail and the user will definitely have permission to write to it

fbs replied:
I had changed it but failed to push the latest version. 

The runtime tests need to be run as root so I'm not sure when write permissions would be an issue. However, I dont see any issues with putting it in /tmp instead of the build dir.

ajor replied:
I was thinking maybe some capabilities would allow bpftrace to run without root, but I don't really know. Also we've been having some problems with snap file permissions recently (not for the tests though, of course)
